packages/fxa-auth-server/lib/zendesk-client.js

```diff
@@ -0,0 +1,39 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+const zendesk = require('node-zendesk');
+
+module.exports = config => {
+  const zendeskClient = zendesk.createClient({
+    username: config.zendesk.username,
+    token: config.zendesk.token,
+    remoteUri: `https://${config.zendesk.subdomain}.zendesk.com/api/v2`,
+    disableGlobalState: true,
+  });
+
+  // Promisify our callback functions, note util.promisify can't be used
```

Comment:
We haven't fully moved away from bluebird promisify (being used here https://github.com/mozilla/fxa/blob/b22d8cdeb47cf4dcd52aa90fc8e21f4eedd24725/packages/fxa-auth-server/lib/cache.js#L10), note sure if we wanted to or not, but you could try using that promisify function.

Reply:
@rfk suggested that as well, it didn't seem to work nicely on the zendesk library though.

Reply:
I'm mildly uneasy about this.

The `promisify` implementation here returns native promises but elsewhere we use bluebird promises. Although they share a common API and can be mixed safely, bluebird promises also support other methods that will be missing if someone accidentally tries to invoke them on one of these. Given that we use some of those methods elsewhere in the code, it seems like a lurking footgun if we return native promises from some places and bluebird promises from others. So at the very least, I think you should return bluebird promises from these wrappers.

But I also think it's preferable to use bluebird's `promisify` over re-implementing our own, if we can. Looking at the code, I presume it was that step of aggregating `req` and `result` to an object, because promises can only resolve to a single value. Bluebird has a `multiArgs` boolean arg for that. It aggregates them to an array instead of an object, but they can still be unpacked to named variables on the way out with array destructuring, e.g.:

```js
const Promise = require('bluebird');

// ...

zendeskClient.createRequest = Promise.promisify(zendeskClient.requests.create, {
  multiArgs: true,
  context: zendeskClient.requests,
});

// ...

const [ , result ] = await zendeskClient.createRequest({ /* ... */ });
```

And if we really wanted to eliminate repetition, we could derive the calls to `promisify` from static data, e.g.:

```js
const wrappedMethods = [
  { name: 'createRequest', context: 'requests', target: 'create' },
  { name: 'showUser', context: 'users', target: 'show' },
  { name: 'updateUser', context: 'users', target: 'update' },
];
wrappedMethods.forEach(method => {
  const { name, context, target } = method;
  zendeskClient[name] = Promise.promisify(zendeskClient[context][target], {
    multiArgs: true,
    context: zendeskClient[context],
  });
});
```

Last thing, I think we prefer `camelCase` names in most JS code. Mixing `snake_case` in some places makes things less uniform across the board, which means APIs become less predictable and intuitive when you're working with them. Neither one is better than the other, I just think there's value in consistency (even though we're not 100% consistent in all places, I realise).

Reply:
I tried both uses of bluebird promisify that you mentioned. Neither works, and results in errors about undefined like this:
```
TypeError: Cannot read property 'requester_id' of undefined
```

Reply:
> ```
> TypeError: Cannot read property 'requester_id' of undefined
> ```

Beats me, let's do it your way! :smile: